File: Week3/Code/vectorize1.py
#!/usr/bin/env python3
"""Python version of vectorize1.R"""
__appname__ = "vectorize1.py"
__author__ = "Joseph Palmer <joseph.palmer18@imperial.ac.uk>"
__version__ = "0.0.1"
__license__ = "License for this code/"
__date__ = "Oct-2018"

## imports ##
import sys
import numpy as np
import time

def SumAllElements(M):
    Dimensions = M.shape
    total = 0
    for i in range(0,Dimensions[0]):
        for j in range(0,Dimensions[-1]):
            total = total + M[i, j]
    return total


# create random matrix
M = np.random.rand(1000, 1000)

# time the looping function
start = time.time()       
SumAllElements(M)
end = time.time()
elapsed = end - start
print("Python SumAllElements function: {}".format(elapsed))

# time the vectorized sum function.
start = time.time()
M.sum()
end = time.time()
elapsed2 = end - start
print("Python sum vectorised function: {}".format(elapsed2))

File: Week3/Code/vectorize1.R
#!/usr/bin/env Rscript
# Author: Joseph Palmer <Joseph.Palmer18@imperial.ac.uk>
# Date: October 2018
# Desc: An example of vectorizing and loops with a speed comparison

# clear environment
rm(list=ls())

# Load required packages #


rm(list=ls())

M <- matrix(runif(1000000),1000,1000)

SumAllElements <- function(M){
  Dimensions <- dim(M)
  Tot <- 0
  for (i in 1:Dimensions[1]){
    for (j in 1:Dimensions[2]){
      Tot <- Tot + M[i,j]
    }
  }
  return (Tot)
}

# Get execution time for loop
start_time <- Sys.time()
a = SumAllElements(M)
end_time <- Sys.time()
cat(paste("R SumAllElements function:", end_time - start_time, "\n"))

# Get execution time for vectorised
start_time <- Sys.time()
a = sum(M)
end_time <- Sys.time()
cat(paste("R vectorised sum function:", end_time - start_time, "\n"))




## This on my computer takes about 1 sec
#print("R SumAllElements function:")
#print(system.time(SumAllElements(M)))
## While this takes about 0.01 sec
#print("R vectorised sum function:")
#print(system.time(sum(M)))
